things/service.go

```diff
@@ -247,13 +241,41 @@ func (ts *thingsService) UpdateThing(ctx context.Context, token string, thing Th
 	return ts.things.Update(ctx, thing)
 }
 
+func (ts *thingsService) ShareThing(ctx context.Context, token, thingID string, policies, userIDs []string) error {
+	res, err := ts.auth.Identify(ctx, &mainflux.Token{Value: token})
+	if err != nil {
+		return errors.Wrap(ErrUnauthorizedAccess, err)
+	}
+
+	if err := ts.authorize(ctx, res.GetId(), thingID, writeRelationKey); err != nil {
+		return err
+	}
+
+	return ts.claimOwnership(ctx, thingID, policies, userIDs)
+}
+
+func (ts *thingsService) claimOwnership(ctx context.Context, thingID string, policies, userIDs []string) error {
+	for _, userID := range userIDs {
+		for _, policy := range policies {
+			apr, err := ts.auth.AddPolicy(ctx, &mainflux.AddPolicyReq{Obj: thingID, Act: policy, Sub: userID})
+			if err != nil {
+				return errors.Wrap(ErrAuthorization, err)
+			}
+			if !apr.GetAuthorized() {
+				return ErrAuthorization
+			}
+		}
+	}
+	return nil
+}
```

Comment:
Wrapping multiple requests in a single method that receives a list is convenient for the user, but it comes at the cost of losing atomic operation: this can fail for any of the provided user IDs. I'd say that this way is OK, but we need to improve error handling. If sharing with some users fails, we need to return it in the error response and add policies for all the other users (so keep iterating even if we fail). The other option is to create a bulk endpoint on the Auth service, which will result in similar behavior (only on the Auth side) unless there is a bulk option provided by Keto.

Reply:
@buraksekili waht is the best way to proceed with this? should we have a meeting on this?

Reply:
I would say `If sharing with some users fails, we need to return it in the error response and add policies for all the other users (so keep iterating even if we fail). `

Reply:
I would say so, as well. But we need to design a nice error response for this.  Something like:
```go
func (ts *thingsService) claimOwnership(ctx context.Context, thingID string, policies, userIDs []string) error {
	var errors Error
        for _, userID := range userIDs {
		for _, policy := range policies {
			apr, err := ts.auth.AddPolicy(ctx, &mainflux.AddPolicyReq{Obj: thingID, Act: policy, Sub: userID})
			if err != nil {
				accumulateError(errors, errors.Wrap(ErrAuthorization, err), thingID, userID)
			}
			if !apr.GetAuthorized() {
				accumulatError(errors, ErrAuthorization, thingID, userID)
			}
		}
	}
	return errors
}
```